server/src-lib/Hasura/GraphQL/Remote/Input.hs

```diff
@@ -0,0 +1,137 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE DisambiguateRecordFields #-}
+
+-- | Input object types.
+
+module Hasura.GraphQL.Remote.Input
+  ( substituteVariables
+  , remoteArgumentsToMap
+  , SubstituteError(..)
+  , RemoteArguments(..)
+  ) where
+
+import           Data.Aeson (FromJSON(..),ToJSON(..))
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Types as Aeson
+import           Data.Foldable
+import           Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HM
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+import           Data.Scientific (floatingOrInteger)
+import           Data.Text (Text)
+import qualified Data.Text as T
+import           Data.Validation
+import qualified Language.GraphQL.Draft.Syntax as GraphQL
+import           Language.Haskell.TH.Syntax (Lift)
+import           Prelude
+
+--------------------------------------------------------------------------------
+-- Type definitions
+
+newtype RemoteArguments =
+  RemoteArguments
+    { getRemoteArguments :: [GraphQL.ObjectFieldG GraphQL.Value]
+    } deriving (Show, Eq, Lift)
+
+instance ToJSON RemoteArguments where
+  toJSON (RemoteArguments fields) = fieldsToObject fields
+
+instance FromJSON RemoteArguments where
+  parseJSON = parseRemoteArguments
+
+-- | An error substituting variables into the argument list.
+data SubstituteError
+  = ValueNotProvided !GraphQL.Variable
+  deriving (Show, Eq)
+
+--------------------------------------------------------------------------------
+-- Operations
+
+-- | Substitute values in the argument list.
+substituteVariables ::
+     Map GraphQL.Variable GraphQL.ValueConst
+  -- ^ Values to use.
+  -> [GraphQL.ObjectFieldG GraphQL.Value]
+  -- ^ A template.
+  -> Validation [SubstituteError] [GraphQL.ObjectFieldG GraphQL.ValueConst]
+substituteVariables values = traverse (traverse go)
+  where
+    go =
+      \case
+        GraphQL.VVariable variable ->
+          case M.lookup variable values of
+            Nothing -> Failure [ValueNotProvided variable]
+            Just valueConst -> pure valueConst
+        GraphQL.VInt int32 -> pure (GraphQL.VCInt int32)
+        GraphQL.VFloat double -> pure (GraphQL.VCFloat double)
+        GraphQL.VString stringValue -> pure (GraphQL.VCString stringValue)
```

Comment:
Are we not using `Hasura.GraphQL.Validate.InputValue`

Reply:
Not for this, as this substitution applies only to the remote arguments template.